### DISCUSSION TITLE
How to properly pass variables between different classes using ManagedDockWindow?

### DISCUSSION THREAD

**Discussion by imbrapy** [pymeasure/pymeasure#967] [Q&A] [answered]
Hi guys! I would need help please...
Let's for now simply refer to the GUI tutorial with the random example which makes use of ManagedDockWindow.
What if I define a variable at the end of the _init_ method in MainWindow, and I want to access it from the execute method in RandomProcedure?
Thanks a lot in advance!

**Comment by msmttchr** [comment]
Well, I guess this is a python question, more than pymeasure specific.
Coming back to your case, you could:
- Set an instance attribute at the end of `__init__`, eg. `self.value=10`
- Add a method to pass a value to `RandomProcedure` eg. `set_value(...)` which record the value passed as instance attribute
- Finally in the `queue` method after `procedure` is assigned, you could call `procedure.set_value(self.value)`

This should work, but, generally speaking, you would like to review whether is a good design choice to proceed like this to solve your problem.

Hope this helps

**Reply by imbrapy** [reply]
Grazie msmttchr, yes sure it works in this way. Indeed, I was just wondering if there is a better design choice provided for pymeasure already. I actually created a lot of customized widgets (doublespinbox, pushbutton, etc.) connected to functions defined in the _init_ method of the GUI class, and I would like to call both widget's values and functions in the procedure class trying also to optimize my script. Since it seems to me that, for instance, inputs defined in the child procedure are managed in a different way, and directly called in the constructor of ManagedDockWindow, I thought I had to implement a similar approach also for my customized widgets and functions, maybe by passing my variables as keyword arguments in the Procedure constructor or using one of the functions defined in the Procedure class or using widget_list. Anyway, I failed so far, and any further suggestion in this sense is more than welcome... Meanwhile, thanks again!

**Comment by msmttchr** [comment]
I think your design choice depends on where the extra widgest you are adding are placed in the GUI area.
If, for example, they are `InputsWidget` then you need to look at
- class `Parameter` and its subclasses
- class `Input` and its subclasses 
- class `InputsWidget` 

to extend them in your application so that your new widgets can be supported.
BTW, doublespinbox already exists and you can use it.

Hope this helps

**Reply by imbrapy** [reply]
Thanks again msmttchr. I know doublespinbox already exists in pymeasure as in the ScientificInput class, and also QLineEdit, QSpinBox, QCheckBox, QComboBox are provided (but not QPushButton). Therefore, what I want to connect one of these input widgets to methods defined in my Procedure class and have access at the same time to the parent attributes from QtWidgets (not only those defined by pymeasure)? For instance, since I have not found radio button, I used BooleanParameter (connected to BooleanInput and QCheckBox) but I am not able to connect to the signal stateChanged() of QCheckBox, only to value of BooleanParameter. And this is the simplest case. Is there an easy way to make some inputs defined in the Procedure class as clickable objects connected to methods defined in the same class? If so, my design choice would definitely change and would be much more compact, as I would be no more obliged to define my variables 2 times by a method which pass them from the Procedure class to the ManagedDockWindow one and viceversa. I found no examples on this, neither in the documentation. If you could help with some advice at this stage, it would be more than welcome!

**Reply by msmttchr** [reply]
I think, you need to use the `ui_class` parameter of the `Parameter` class.
See attached file which is modified version of the example `gui.py`.
[gui.zip](https://github.com/pymeasure/pymeasure/files/13047126/gui.zip)

**Reply by imbrapy** [reply]
Yes, this seems what I was looking for... I'll play with it. Thanks!

**Reply by imbrapy** [reply]
[gui_imbrapy.zip](https://github.com/pymeasure/pymeasure/files/13179812/gui_imbrapy.zip)
Could you please explain to me why (in this slightly modified example with BooleanInput) the variable working works and the variable notworking does not? All the 2 refer to the same BooleanParameter, whose default value is set as False; the fomer prints True, the latter gives _AttributeError: 'bool' object has no attribute 'is_set'_. I would simply need to interrogate continuosly during the execute cycle if the correction checkbox is selected or not by the user, but it seems to me harder than expected at this stage 🥲

**Reply by msmttchr** [reply]
Ah ah, you hit the obscure part!
Let me try to explain how this working, in my understanding.

First of all, here we are talking about python and not pymeasure.

1) I guess you know what are class attributes and instance attributes
2) Experiment parameters are defined as class attributes that can generate gui input widgets.
3) When the class is instantiated all the values from input widgets are copied to instance attributes whose name is exactly as the corresponding class attribute

In your case, the parameter correction_on_off is declared as class attributes of type BooleanParameter.
Indeed your print of the variable working is done at class declaration level and so has the expected type.

After instantiation, in the startup method, this has been changed to an a bool type capturing the value in the corresponding input widget.

To prove this, you could replace your print with something like
`print (correction_on_off.__class__)`

and 

`print (self.correction_on_off.__class__)`

If you understand why different values are printed, you are done, otherwise keep trying :)

This is a feature of python that is very handy, but difficult to explain (at least for me) and to understand.

Remember that understanding point 1) above is key.

**Reply by imbrapy** [reply]
Dear msmttchr, sorry for the very late reply but I kept trying on this project only in my free time (very hard times...). Anyway, of course, thanks a lot for your suggestions.
I better understood how class and instance attributes work in python, anyway, still some issues to play with them in pymeasure.
First thing I tried has been to move my `correction_on_off` BooleanParameter from class attributes to the TestProcedure class constructor as instance attribute: I obtain a `KeyError: 'correction_on_off'` in inputs_widget.py:58 related to the `inputs` list of the GUI which I cannot solve. Then, since the only way I know to visualize the `correction_on_off` checkbox is to keep it together with the other parameters, I created an instance attribute inside the ui_class associated with the boolean parameter, i.e. `MyBooleanInput`. This works well and, once the checkbox is clicked, it correctly updates the True/False value of the `self.instance_attr_input` variable (but I had to write my own if else statement). Therefore, my idea was to create another instance attribute `self.instance_attr_procedure` in the Procedure class with the ForLoop where I want to print it, and to link `self.instance_attr_input` to `self.instance_attr_procedure` somehow. I realized 2 things:
1. I'm quite confused when calling a particular instance in pymeasure (above all, the instance of the running Testprocedure is procedure right?)
2. maybe my approach is wrong/redundant, since I will add in the future many other inputs

I add a working py only able to print `<function BooleanInput.value at 0x000002178B613880>` where I would like to, not the real time True/False value of the checkbox.
As usual, many thanks in advance for helping.
Best
[gui_imbrapy_2.zip](https://github.com/pymeasure/pymeasure/files/14053152/gui_imbrapy_2.zip)

**Comment by msmttchr** [comment, marked as answer]
@imbrapy, long delay I hardly remember this :).
From your program, I see that every time the `TestProcedure` is instantiated, you create an instance attribute called `instance_attr_procedure` whose value is `MyBooleanInput.value`  (method value of `MyBooleanInput` class, constant value).

I am also confused on what you are trying to do, sorry.

Let me clarify one point regarding the GUI behavior and parameters handling.
The GUI is supposed to work in the following way:
- User defines the inputs parameters
- User clicks Queue: in this moment a new instance of the `TestProcedure` is created taking a snapshot of the current input values. From now on any change in the input area will not affect in anyway the queued instance of `TestProcedure`.

If you are trying to change input parameters after experiment is queue or while experiment is running, this is simply not possible.
On the other hand, you can change the visual presentation of results, by acting on the relevant command on the right area of the GUI, e.g. you can change color of the curve.

For example, if you run an experiment and you would like to view results with or without correction applied, you probably need to implement your own GUI widget and implement the command to enable/disable correction (see https://pymeasure.readthedocs.io/en/latest/tutorial/graphical.html#defining-your-own-managedwindow-s-widgets)

Sorry, I am not sure if I am helpful.

**Reply by imbrapy** [reply]
@msmttchr , no, you've been a big help!... by just saying that what I'm trying to do now is impossible :)
As you can get from my previous posts, I already tried to implement my own GUI widgets in different ways. I skipped the option of using `widget_list` since there is no checkbox (or doublespinbox, combobox, pushbutton, etc) among the widgets already provided by pymeasure, but only 5 of them which inherit from `TabWidget`. I was only able to make work `TableWidget` and `LogWidget` so far, but If I well understood I need to write something like `MyCheckboxWidget` in my case. I hope it will not be too difficult, any further example is more than welcome of course! If you don't mind, I will leave the discussion unanswered to let me bother you from time to time until I solve this issue... Thanks again @msmttchr !
